mailpoet/lib/Validator/Schema.php

```diff
@@ -0,0 +1,102 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Validator;
+
+use MailPoet\InvalidStateException;
+
+use function json_encode;
+use function rest_get_allowed_schema_keywords;
+
+abstract class Schema {
+  protected $schema = [];
+
+  /** @return static */
+  public function nullable() {
+    $type = $this->schema['type'] ?? ['null'];
+    return $this->updateSchemaProperty('type', is_array($type) ? $type : [$type, 'null']);
+  }
+
+  /** @return static */
+  public function nonNullable() {
+    $type = $this->schema['type'] ?? null;
+    return $type === null
+      ? $this->unsetSchemaProperty('type')
+      : $this->updateSchemaProperty('type', is_array($type) ? $type[0] : $type);
+  }
+
+  /** @return static */
+  public function required() {
+    return $this->updateSchemaProperty('required', true);
+  }
+
+  /** @return static */
+  public function optional() {
+    return $this->unsetSchemaProperty('required');
+  }
+
+  /** @return static */
+  public function title(?string $title) {
+    return $title === null
+      ? $this->unsetSchemaProperty('title')
+      : $this->updateSchemaProperty('title', $title);
+  }
+
+  /** @return static */
+  public function description(?string $description) {
+    return $description === null
+      ? $this->unsetSchemaProperty('description')
+      : $this->updateSchemaProperty('description', $description);
+  }
+
+  /** @return static */
+  public function default($default) {
+    return $default === null
+      ? $this->unsetSchemaProperty('default')
+      : $this->updateSchemaProperty('default', $default);
+  }
+
+  /** @return static */
+  public function field(string $name, $value) {
+    if (in_array($name, $this->getReservedKeywords(), true)) {
+      throw new InvalidStateException("Field name '$name' is reserved");
+    }
+    return $value === null
+      ? $this->unsetSchemaProperty($name)
+      : $this->updateSchemaProperty($name, $value);
+  }
+
+  public function toArray(): array {
+    return $this->schema;
+  }
```

Comment:
You're right, it doesn't seem to be used though. So I don't know... should we add it somehow? The thing is we can't make it being added automatically because the schema factories don't know which one of them is root 🤔